src/plugins/intel_cpu/src/nodes/kernels/gather_uni_kernel.cpp

```diff
@@ -10,1036 +10,1157 @@ using namespace dnnl::impl::cpu;
-                add(rSpecIdxAndAfterAxIterB, idxElPerVec * jcp.dataTypeSize);
-                cmp(rSpecIdxAndAfterAxIterB, rSpecIdxAndAfterAxSizeB);
-                jl(lBeforeAxStepEnd1, T_NEAR);
-                    sub(rSpecIdxAndAfterAxIterB, rSpecIdxAndAfterAxSizeB);
-                cmp(rSpecIdxAndAfterAxIterB, 0);
-                jne(lBeforeAxStepEnd1, T_NEAR);
-                    uni_vpaddd(vmmSrcBeforeAxisSumB, vmmSrcBeforeAxisSumB, vmmAxisAndAfterAxisSizeB);
-                L(lBeforeAxStepEnd1);
+                this->uni_vmovups(this->ptr[this->regDst], vAux0);
             }
+
+            this->jmp(lDstIdxLoop1, this->T_NEAR);
         }
     }
+    this->L(lTail);
+    this->tail(shiftCalculator, true);
+}
 
-    vpcmpeqd(kDstMask, vAux0, vAux0);
-    if (jcp.batchDims > 0lu) {
-        // Calculate indices batch sum.
-        uni_vcvtdq2ps(vAux0, vAux1);
-        uni_vcvtdq2ps(vDstShifts, vmmSrcAfterBatchSizeB);
-        uni_vdivps(vAux0, vAux0, vDstShifts);
-        uni_vroundps(vAux0, vAux0, 0x1);
-        uni_vcvtps2dq(vAux0, vAux0);
+template<x64::cpu_isa_t isa, DataTypeSize S, AfterAxisCase C, Approach A>
+void jitGatherKernelForStaticShapes<isa, S, C, A>::generateDynamicitySpecific() {
+    this->uploadParamPtrWithVpbroadcastd(this->vmmSpecIdxSizeB, GET_OFF(specIndicesSize));
+    this->uni_vpslld(this->vmmSpecIdxSizeB, this->vmmSpecIdxSizeB, this->idxTypeShift); // multiply by indexes type size.
 
-        uni_vpmulld(vAux0, vAux0, vmmSpecIdxSizeB);
-        uni_vpaddd(vAux0, vAux0, vmmSpecIdxB);
+    this->uploadParamPtrWithVmovups(this->vmmSpecIdxB, GET_OFF(specIdxB));
 
-        uniVpGatherDd(vDstShifts, ptr[regIndices + vAux0], kDstMask);
-    } else {
-        uniVpGatherDd(vDstShifts, ptr[regIndices + vmmSpecIdxB], kDstMask);
+    if (this->beforeAxisSize != 1lu) {
+        this->uploadParamPtrWithVmovups(this->vmmSrcBeforeAxisSumB, GET_OFF(dataBeforeAxisSumB));
     }
+    shiftCalculator.allocateRegisters(*this);
+    shiftCalculator.uploadParamsForApproachSpecific(*this);
+    this->process(shiftCalculator);
+}
 
-    normalizeRawIndices(vDstShifts, kDstMask, kAuxMask0);
+template<x64::cpu_isa_t isa, DataTypeSize S>
+void jitGatherKernelForDynamicShapes<isa, S>::generateDynamicitySpecific() {
+    typename jitGatherKernelForIsa<isa>::template ShiftCalculatorImpl<ElementwiseCase, Long> elementwiseLong;
+    typename jitGatherKernelForIsa<isa>::template ShiftCalculatorImpl<ElementwiseCase, Short> elementwiseShort;
+    typename jitGatherKernelForIsa<isa>::template ShiftCalculatorImpl<BlockedCase, Short> blockedShort;
+    elementwiseLong.allocateRegisters(*this);
 
-    if (jcp.afterAxisSize != 1lu) {
-        vpmulld(vDstShifts, vDstShifts, vmmAfterAxisSize);
-        uni_vpaddd(vDstShifts, vDstShifts, vmmAfterAxisIdxB);
+    static const unsigned incVec[16] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15};
+    this->uploadParamPtrWithVpbroadcastd(this->vmmSpecIdxB, GET_OFF(start));
```

Comment:
Why do you use *this->* everywhere?

Reply:
@nshchego It is because these members of the base class are dependent names. See [here](https://stackoverflow.com/questions/7908248/in-a-templated-derived-class-why-do-i-need-to-qualify-base-class-member-names-w) and [here](https://en.cppreference.com/w/cpp/language/dependent_name).

Reply:
Oh, I see... There are a lot of templates now. Please reduce number of templates and use them just if really needed. They produce a lot of code that do the same work.

Reply:
I reduced the use of templates to avoid generating the same code.